REGRESSION(291757@main)[WPE] g_warning_once compile failure with clang

lauromoura opened the pull request:
#### 8cf192df965d90c8bbafaf0629669429f6f6d45a
<pre>
REGRESSION(291757@main)[WPE] g_warning_once compile failure with clang
<a href="https://bugs.webkit.org/show_bug.cgi?id=289343">https://bugs.webkit.org/show_bug.cgi?id=289343</a>

Unreviewed build fix.

g_warning_once uses g_atomic_int_compare_and_exchange, which seems to
have a history of issues with clang (e.g.
<a href="https://gitlab.gnome.org/GNOME/glib/-/issues/2625)">https://gitlab.gnome.org/GNOME/glib/-/issues/2625)</a>

This commit replaces g_warning_once with plain g_warning, as the warning
is not on a hot path.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(wpeDisplayDRMSetup):

Canonical link: <a href="https://commits.webkit.org/291781@main">https://commits.webkit.org/291781@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7860343dcd04adcda0c6259a4ff4de51365db070

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94012 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99020 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44539 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22029 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71740 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29088 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10319 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52079 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10001 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43855 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2658 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101061 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21064 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21316 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80898 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80109 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19942 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24662 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2018 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14226 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21048 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26226 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20735 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24195 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22476 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -330,8 +330,11 @@ static gboolean wpeDisplayDRMSetup(WPEDisplayDRM* displayDRM, const char* device
+            static std::atomic<bool> warnedOnce = false;
+            if (!warnedOnce.exchange(true))
+                g_warning("Invalid WPE_DRM_SCALE='%*s' value, or out of bounds.", static_cast<int>(scaleString.span8().size()), scaleString.span8().data());
+        }
```
I would just use `g_warning` without the atomic. It' no big deal to have a line printed each time a web view is created, because that's not an operation that would be happening the whole time, only seldomly.

```suggestion
        g_warning("Invalid WPE_DRM_SCALE='%*s' value, or out of bounds.", ```

mcatanzaro replied:
You don't expect people will try to use WPE to build a tabbed browser? I saw rumblings that Emacs developers were considering switching from WebKitGTK to WPE, for example. Not sure how serious that is, though.

Notably, I also notice that this build failure was already fixed in GLib several years ago.

mcatanzaro replied:
(But this code doesn't look like it's intended to be multithreaded, so I agree that using an atomic is overkill.)

aperezdc replied:
Yes, we expect people to use multiple web views with WPE these days, but IMO one warning for each web view that gets created is not such a big deal, just a tiny little itsy bit of log spamming. I'd rather not bring in additional code and just use `g_warning` until our GLib requirement is high enough to be able to use `g_warning_once` 😉

aperezdc commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -348,8 +351,11 @@ static gboolean wpeDisplayDRMSetup(WPEDisplayDRM* displayDRM, const char* device
+        static std::atomic<bool> warnedOnce = false;
+        if (!warnedOnce.exchange(true))
+            g_warning("Ignoring out of bounds WPE_SETTING_DRM_SCALE value '%.2f'.", scaleFromSettings);
+    }
```
Same here:

```suggestion
    g_warning("Ignoring out of bounds WPE_SETTING_DRM_SCALE value '%.2f'.", scaleFromSettings);
```

aperezdc reviewed (approved):
LGTM, with the suggested changes applied before landing.

lauromoura committed:
REGRESSION(291757@main)[WPE] g_warning_once compile failure with clang
https://bugs.webkit.org/show_bug.cgi?id=289343

Unreviewed build fix.

g_warning_once uses g_atomic_int_compare_and_exchange, which seems to
have a history of issues with clang (e.g.
https://gitlab.gnome.org/GNOME/glib/-/issues/2625)

This commit replaces g_warning_once with plain g_warning, as the warning
is not on a hot path.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(wpeDisplayDRMSetup):

Canonical link: https://commits.webkit.org/291781@main
